src/Database/LSMTree/Internal/Snapshot.hs

```diff
@@ -485,23 +483,52 @@ fromSnapLevels reg hfs hbio conf uc resolve dir (SnapLevels levels) =
          LevelNo
       -> SnapIncomingRun (Ref (Run m h))
       -> m (IncomingRun m h)
-    fromSnapIncomingRun ln (SnapSingleRun run) =
-        newIncomingSingleRun tr ln run
-
-    fromSnapIncomingRun ln (SnapMergingRun mpfl nr md _nc
-                                           (SnapCompletedMerge r)) =
-      newIncomingCompletedMergingRun tr conf ln mpfl nr md r
-
-    fromSnapIncomingRun ln (SnapMergingRun mpfl _nr _md nc
-                                           (SnapOngoingMerge rs mt)) = do
-      bracketOnError (newIncomingMergingRun tr hfs hbio dir uc
-                                  conf resolve
-                                  mpfl mt ln rs) releaseIncomingRun $ \ir -> do
+    fromSnapIncomingRun _ln (SnapSingleRun run) =
+        newIncomingSingleRun run
+
+    fromSnapIncomingRun ln (SnapMergingRun mergePolicy nominalDebt
+                                           nominalCredits smrs) = do
+      mr <- fromSnapMergingRunState hfs hbio conf uc resolve dir smrs
+      ir <- newIncomingMergingRun mergePolicy nominalDebt mr
+      -- This will set the correct nominal credits, but it will not do any more
+      -- merging work because fromSnapMergingRunState already supplies all the
+      -- merging credits already.
+      supplyCreditsIncomingRun conf ln ir nominalCredits
```

Comment:
Just to make sure I understand it correctly:

`it will not do any more merging work` is based on the invariant that the snapshotted merging run's physical/merge credits were at least as high as dictated by the incoming run's nominal credit. If that assumption wasn't true for some reason, we would fix that here by doing some merging work. 

Is that correct?

Reply:
That is correct.